py/desispec/zcatalog.py

```diff
@@ -1,15 +1,429 @@
+    ## Add the required columns or select a few of them
+    ## Adding all the tables into a single list
+    tables = []
+    
+    ## Looping through the different zcatalogs and fixing their columns
+    for filename in zcat:
+        arr = filename.split('-')
+        ## Filename can be used to get the survey and program of the given redshift catalog
+        survey = arr[1]
+        program = arr[2].split('.')[0]
+        
+        ## Sanity check -- can be removed later
+        print (survey, program)
+        
+        ## Create variable names for every table
+        table_name = f'{survey}_{program}'
+        
+        ## The function 'fix_tables' adds the required columns 
+        ## Returns the final table with all or selected columns
+        vars()[table_name] = fix_tables(survey, program, spec_release = spec_release,\
+                                        version = version, all_columns = all_columns, \
+                                        columns_list = columns_list)
+        
+        ## Appending the tables to the list
+        tables.append(vars()[table_name])
+        
+    
+    ## Stacking all the tables into a final table
+    tab = vstack(tables)
+    
+    ## Selecting primary spectra for the whole combined ZCATALOG
+    ## For SV, it selects the best spectrum including cmx+special+sv1+sv2+sv3
+    ## For Main, it selects the best spectrum for main+special
+    nspec, specprim = find_primary_spectra(tab)
+    
+    ## Replacing the existing 'ZCAT_NSPEC' and 'ZCAT_PRIMARY'
+    ## If all_columns = False, and user-list does not contain this column -
+    ## these columns will be added 
+    tab['ZCAT_NSPEC'] = nspec
+    tab['ZCAT_PRIMARY'] = specprim
+    
+    ## Adding empty columns for SV|MAIN NSPEC and PRIMARY
+    col1 = Column(np.array([0]*len(tab)), name = f'{survey_name.upper()}_NSPEC', dtype = '>i8')
+    col2 = Column(np.array([0]*len(tab)), name = f'{survey_name.upper()}_PRIMARY', dtype = 'bool')
+    tab.add_columns([col1, col2], [-4, -3])
+    
+    ## Selecting primary spectra for targets within just SV or MAIN depending on the survey_name
+    ## For SV, it selects the primary spectra out of SV1+SV2+SV3 for every individual TARGETID
+    ## Ignores cmx+special in SV
+    ## For Main, it selects the primary spectra just for 'main' and ignores 'special'
+    is_survey = np.char.startswith(tab['SURVEY'].astype(str).data, survey_name)
+    nspec, specprim = find_primary_spectra(tab[is_survey])
+    tab[f'{survey_name.upper()}_NSPEC'][is_survey] = nspec
+    tab[f'{survey_name.upper()}_PRIMARY'][is_survey] = specprim
+    
+    ## For convenience, sort by SURVEY, PROGRAM, and (HEALPIX or TILEID) 
+    if (version == 'zpix'):
+        tab.sort(['SURVEY', 'PROGRAM', 'HEALPIX'])
+    else:
+        tab.sort(['SURVEY', 'PROGRAM', 'TILEID', 'LASTNIGHT'])
+    
+    ## Add the final table as a HDU
+    hdu0 = fits.PrimaryHDU()
+    hdu1 = fits.BinTableHDU(tab)
```

Comment:
I think this is where the column comments are getting lost (maybe), and I don't think you need to go via HDUList etc. to write a single table.  Consider:
```
tab.meta['EXTNAME'] = 'ZCATALOG'
tab.write(output_filename, overwrite=True)
```
otherwise try using `fits.convenience.table_to_hdu(tab)` which may preserve the column comments.

Reply:
Both of these suggestions do not preserve the column comments. I changed it to your first suggestion.